src/main/routes/receiver.ts

```diff
@@ -1,49 +1,209 @@
 import * as express from 'express'
-import * as config from 'config'
 
 import { Paths as AppPaths } from 'app/paths'
 import { Paths as DashboardPaths } from 'dashboard/paths'
 import { Paths as ClaimPaths } from 'claim/paths'
 import { Paths as ResponsePaths } from 'response/paths'
+import { ErrorPaths, Paths as FirstContactPaths } from 'first-contact/paths'
 import ClaimStoreClient from 'app/claims/claimStoreClient'
-
-import * as Cookies from 'cookies'
 import User from 'app/idam/user'
 import { ErrorHandling } from 'common/errorHandling'
 import Claim from 'claims/models/claim'
+import * as toBoolean from 'to-boolean'
+import * as Cookies from 'cookies'
+import { AuthToken } from 'idam/authToken'
+import * as config from 'config'
+import IdamClient from 'app/idam/idamClient'
+import { buildURL } from 'utils/callbackBuilder'
+import DraftClaim from 'drafts/models/draftClaim'
+import JwtExtractor from 'idam/jwtExtractor'
+import { RoutablePath } from 'common/router/routablePath'
+import { DraftStoreClientFactory } from 'common/draft/draftStoreClientFactory'
+import DraftStoreClient from 'common/draft/draftStoreClient'
+import { Draft } from 'models/draft'
+import { hasTokenExpired } from 'idam/authorizationMiddleware'
+import { AuthenticationRedirectFactory } from 'utils/AuthenticationRedirectFactory'
+import { AuthenticationRedirect } from 'utils/authenticationRedirect'
+
+const logger = require('@hmcts/nodejs-logging').getLogger('router/receiver')
+const useOauth = toBoolean(config.get<boolean>('featureToggles.idamOauth'))
+const sessionCookie = config.get<string>('session.cookieName')
+
+const authenticationRedirect: AuthenticationRedirect = AuthenticationRedirectFactory.get()
+
+async function getOAuthAccessToken (req: express.Request, receiver: RoutablePath): Promise<string> {
+  if (req.query.state !== authenticationRedirect.getStateCookie(req)) {
+    throw new Error('Invalid state')
+  }
+  const authToken: AuthToken = await IdamClient.exchangeCode(
+    req.query.code,
+    buildURL(req, receiver.uri)
+  )
+  return authToken.accessToken
+}
+
+async function getAuthToken (req: express.Request,
+                             receiver: RoutablePath = AppPaths.receiver,
+                             checkCookie = true): Promise<string> {
+  let authenticationToken
+  if (!useOauth && req.query.jwt) {
+    authenticationToken = req.query.jwt
+  } else if (useOauth && req.query.code) {
+    authenticationToken = await getOAuthAccessToken(req, receiver)
+  } else if (checkCookie) {
+    authenticationToken = JwtExtractor.extract(req)
+  }
+  return authenticationToken
+}
+
+function isDefendantFirstContactPinLogin (req: express.Request): boolean {
+  return useOauth && req.query && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```

Comment:
are you sure we need to have this chain of conditions:

```
req.query && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```

and not just:
```
req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```

Reply:
tested and its fine with the simpler version

Reply:
I was testing with the feature toggle off like a dummy, good thing there was a test,
the req.query check isn't needed, all the others are though
```
return useOauth && req.query.state && req.query.state.match(/[0-9]{3}MC[0-9]{3}/)
```